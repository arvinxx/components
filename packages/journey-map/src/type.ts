/**
 * 阶段
 */
export interface Step<T = string> {
  /**
   * 阶段id
   * @TJS-type string
   */
  id: T;
  /**
   * 阶段名称
   */
  name: string;
  /**
   * 该阶段的颜色
   */
  color?: string;
}

/**
 * 用户旅程地图数据
 */
export interface JourneyMapData<T extends string = string> {
  /**
   * 大的步骤
   */
  steps: Step<T>[];
  /**
   * 用户行为
   */
  actions: Record<T, UserAction[]>;
}

/**
 * 用户行为
 */
export interface UserAction {
  /**
   * 行为名称
   */
  title: string;
  /**
   * 说明
   */
  description?: string;

  /**
   * 有关的想法
   */
  thoughts?: string[];
  /**
   * 情绪得分
   */
  emotion: number;
  /**
   * 痛点
   */
  painSpot?: string[];
}

/**
 * 解析后的 YML 数据
 */
export interface JourneyMapYML<T extends string = string> {
  /**
   * 步骤
   */
  steps: Step<T>[];
  /**
   * 用户行为
   */
  actions: Record<T, Record<T, number>[]>;
}
