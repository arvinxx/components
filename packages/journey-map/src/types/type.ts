/**
 * 阶段
 */
export interface Stage<T = string> {
  /**
   * 阶段id
   * @TJS-type string
   */
  id: T;
  /**
   * 阶段名称
   */
  name: string;
  /**
   * 该阶段的颜色
   */
  color?: string;
}

/**
 * 用户旅程地图数据
 */
export interface JourneyMapData<T extends string = string> {
  /**
   * 大的步骤
   */
  stages: Stage<T>[];
  /**
   * 用户行为
   */
  actions: Record<T, UserAction[]>;
}

/**
 * 用户行为类型
 */
export interface UserAction {
  /**
   * 行为名称
   */
  name: string;
  /**
   * 情绪得分
   * @minimum -2
   * @maximum 2
   * @TJS-type integer
   */
  emotion?: number;

  /**
   * 说明
   */
  description?: string;

  /**
   * 想法
   */
  thoughts?: string[];

  /**
   * 痛点
   */
  painSpot?: string[];
}
