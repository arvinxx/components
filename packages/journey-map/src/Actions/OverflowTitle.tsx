import type { FC } from 'react';
import React, { useRef } from 'react';
import { Tooltip } from 'antd';
import { useOverflow } from 'use-overflow';

interface OverflowTitleProps {
  title: string;
}
const OverflowTitle: FC<OverflowTitleProps> = ({ title }) => {
  const textRef = useRef(null);
  const { refYOverflowing } = useOverflow(textRef);
  return (
    <div
      ref={textRef}
      style={{
        overflow: 'hidden',
        whiteSpace: 'nowrap',
        textOverflow: 'ellipsis',
      }}
    >
      {refYOverflowing ? (
        <Tooltip title={title}>{title}</Tooltip>
      ) : (
        <>{title}</>
      )}
    </div>
  );
};

export default OverflowTitle;
