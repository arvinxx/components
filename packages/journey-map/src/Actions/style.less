.avx-journey-map-actions {
  &-container {
    display: flex;
    flex: 1;
  }

  &-content {
    display: flex;
    flex: 1;
  }

  &-action {
    position: relative;
    display: flex;
    flex-wrap: nowrap;
    justify-content: center;
    margin-top: 8px;
    border-top: 3px solid transparent;

    &:first-child {
      justify-content: flex-start;
      margin-left: 8px;
    }

    &:last-child {
      justify-content: flex-end;
      margin-right: 8px;
    }

    &-point {
      position: absolute;
      top: -6px;
      width: 10px;
      height: 10px;
      border-radius: 10px;
    }

    &-text {
      margin: 0;
      margin-top: 12px;
      color: rgba(0, 0, 0, 0.45);
      font-size: 12px;
      cursor: pointer;
      writing-mode: vertical-lr;
    }
  }
}
