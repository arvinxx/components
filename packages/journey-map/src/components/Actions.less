.avx-journey-map-actions {
  &-container {
    display: flex;
    height: 120px;
    margin-top: 24px;
  }

  &-action {
    position: relative;
    display: flex;
    flex-wrap: nowrap;
    justify-content: center;
    margin-top: 8px;
    border-top: 3px solid transparent;

    &-point {
      position: absolute;
      top: -6px;
      width: 10px;
      height: 10px;
      border-radius: 10px;
    }

    &-text {
      margin: 0;
      margin-top: 12px;
      color: rgba(0, 0, 0, 0.45);
      font-size: 12px;
      cursor: pointer;
      writing-mode: vertical-lr;
    }
  }
}

.avx-journey-map-flow-content {
  > div:first-child {
    .avx-journey-map-actions-container {
      .avx-journey-map-actions-action {
        &:first-child {
          justify-content: flex-start;
        }
      }
    }
  }

  > div:last-child {
    .avx-journey-map-actions-container {
      .avx-journey-map-actions-action {
        &:last-child {
          justify-content: flex-end;
        }
      }
    }
  }
}
