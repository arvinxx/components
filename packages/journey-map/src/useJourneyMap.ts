import useMergeValue from 'use-merge-value';
import { useEffect, useState } from 'react';

import { getServiceToken, YMLToJSON } from './utils';
import type { YMLJourneyMapData, JourneyMapData } from './types';

interface StoreParams {
  data: JourneyMapData | string;
  onChange?: (data: JourneyMapData) => void;
  title?: string;
}

export const useJourneyMap = ({
  data,
  onChange,
  title: initTitle,
}: StoreParams) => {
  const defaultJourneyMap: JourneyMapData = { stages: [], actions: {} };

  const hook = useMergeValue<JourneyMapData>(
    typeof data === 'string' ? defaultJourneyMap : data,
    { onChange },
  );

  const [title, setTitle] = useState<string>(initTitle);

  const [store, setStore] = hook;
  /**
   * 从 YML URL 中解析 JSON
   * @param url
   */
  const fetchYMLToJSON = async (url: string): Promise<YMLJourneyMapData> => {
    const res = await fetch(url);
    const yml = await res.text();

    return YMLToJSON(yml);
  };

  useEffect(() => {
    if (typeof data === 'string' && data.startsWith('http')) {
      fetchYMLToJSON(data).then((result) => {
        const { stages, actions } = result;
        setTitle(result.title);
        setStore({ stages, actions });
      });
    }
  }, [data]);

  const actionList = Object.values(store.actions).flat();

  return { store, actionList, actionLength: actionList.length, title };
};

export const JourneyMapStore = getServiceToken(useJourneyMap);
