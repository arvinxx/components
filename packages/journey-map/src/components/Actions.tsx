import type { FC } from 'react';
import React from 'react';

import type { Stage, JourneyMapData } from '../types';

import './Actions.less';

interface UserActionProps {
  step: Stage;
  stepIndex: number;
  stepLength: number;
  actions: JourneyMapData['actions'];
  color: string;
}
const Actions: FC<UserActionProps> = ({
  step,
  stepIndex,
  actions,
  stepLength,
  color,
}) => {
  const actionList = actions[step.id];

  const isFirstStep = stepIndex === 0;
  const isLastStep = stepIndex === stepLength - 1;

  const isCenterStep = !isFirstStep && !isLastStep;

  return (
    <div className="avx-journey-map-actions-container">
      {actionList.map((action, index) => {
        const n = actionList.length;

        const isFirst = index === 0;
        const isLast = index === n - 1;

        const calcLength = () => {
          const centerLength = 100 / n; // 正常中间的节点
          const firstOrLastLength = 100 / (2 * n - 1);

          // 中间节点则使用正常长度
          if (isCenterStep) return centerLength;

          // 如果是第一个节点或最后一个节点 使用单位长度
          if ((isFirstStep && isFirst) || (isLastStep && isLast))
            return firstOrLastLength;

          return firstOrLastLength * 2;
        };

        return (
          <div
            key={action.name}
            className="avx-journey-map-actions-action"
            style={{
              strokeLinecap: 'round',
              width: `${calcLength()}%`,
              marginLeft: isFirstStep && isFirst ? 8 : 0,
              marginRight: isLastStep && isLast ? 8 : 0,
              borderTopColor: color,
            }}
          >
            <div
              className="avx-journey-map-actions-action-point"
              style={{ background: color }}
            />
            <div className="avx-journey-map-actions-action-text">
              {action.name}
            </div>
          </div>
        );
      })}
    </div>
  );
};

export default Actions;
