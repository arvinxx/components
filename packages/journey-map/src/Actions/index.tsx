import type { FC } from 'react';
import React, { useContext } from 'react';
import { blue } from '@ant-design/colors';

import { Section, OverflowTitle } from '../components';

import { JourneyMapStore } from '../useJourneyMap';
import { calcActionLength } from '../utils';

import './style.less';
import { UserAction } from '../types';

const Actions: FC = () => {
  const { store, actionList } = useContext(JourneyMapStore);
  const { actions, stages } = store;

  const getColor = (action: UserAction) => {
    const { name } = action;
    let color: string = blue[0];

    if (action.color) return action.color;

    stages.forEach((stage) => {
      if (
        stage.color &&
        actions[stage.id].findIndex((a) => a.name === name) > -1
      )
        color = stage.color;
    });

    return color;
  };

  return (
    <div className="avx-journey-map-actions-container" style={{ height: 140 }}>
      <Section height={140}>用户行为</Section>
      <div className="avx-journey-map-actions-content">
        {actionList.map((action, index) => {
          const color = getColor(action);

          return (
            <div
              key={action.name}
              className="avx-journey-map-actions-action"
              style={{
                strokeLinecap: 'round',
                width: `${calcActionLength(actionList.length, index)}%`,
                borderTopColor: color,
              }}
            >
              <div
                className="avx-journey-map-actions-action-point"
                style={{ background: color }}
              />
              <div className="avx-journey-map-actions-action-text">
                <OverflowTitle direction={'y'} title={action.name} />
              </div>
            </div>
          );
        })}
      </div>
    </div>
  );
};

export default Actions;
