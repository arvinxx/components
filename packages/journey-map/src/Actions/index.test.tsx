import React from 'react';
import { render } from '@testing-library/react';

import Actions from './index';
import { JourneyMapStore, useJourneyMap } from '../useJourneyMap';

import type { FC } from 'react';
import type { UserAction } from '../types';

const Demo: FC<{ actions: Record<string, UserAction[]> }> = ({ actions }) => {
  const value = useJourneyMap({
    data: {
      actions,
      stages: [
        { id: '1', name: '1' },
        { id: '2', name: '2', color: 'blue' },
      ],
    },
  });
  return (
    <JourneyMapStore.Provider value={value}>
      <Actions />
    </JourneyMapStore.Provider>
  );
};

describe('Actions', () => {
  it('默认状态', () => {
    const { container } = render(
      <Demo
        actions={{
          '1': [{ name: '1', color: 'red' }],
          '2': [{ name: '2' }],
        }}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
