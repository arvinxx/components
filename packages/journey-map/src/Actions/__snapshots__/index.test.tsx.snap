// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions 默认状态 1`] = `
<div>
  <div
    class="avx-journey-map-actions-container"
    style="height: 140px;"
  >
    <div
      class="avx-journey-map-section"
      style="height: 140px;"
    >
      <div
        class="avx-journey-map-section-text"
      >
        用户行为
      </div>
    </div>
    <div
      class="avx-journey-map-actions-content"
    >
      <div
        class="avx-journey-map-actions-action"
        style="stroke-linecap: round; width: 50%; border-top-color: red;"
      >
        <div
          class="avx-journey-map-actions-action-point"
          style="background: red;"
        />
        <div
          class="avx-journey-map-actions-action-text"
        >
          <div
            style="overflow: hidden; white-space: nowrap; text-overflow: ellipsis;"
          >
            1
          </div>
        </div>
      </div>
      <div
        class="avx-journey-map-actions-action"
        style="stroke-linecap: round; width: 50%; border-top-color: blue;"
      >
        <div
          class="avx-journey-map-actions-action-point"
          style="background: blue;"
        />
        <div
          class="avx-journey-map-actions-action-text"
        >
          <div
            style="overflow: hidden; white-space: nowrap; text-overflow: ellipsis;"
          >
            2
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
